### PR TITLE
[middleware] page draft

### PR DESCRIPTION
**Pull request by sarah11918**
First draft of an experimental middleware page! Something we can start to look at!

cc @ematipico

### CODE REVIEW

File: src/content/docs/en/guides/middleware.mdx [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+## Basic Usage
+
+1. Create `src/middleware.js|ts` (Alternatively, you can create `src/middleware/index.js|ts`.)
+
+2. Inside this file, export an [`onRequest()`](#onrequest) function that modifies a [`locals`](#locals) object. This must not be a default export.
+
+```js title="src/middleware.js"
+export function onRequest ({ locals, request }, resolve) {
+	// locals = {}
```
Comment (matthewp) [inline src/content/docs/en/guides/middleware.mdx:43]: You can't set the value of `locals` here. You can add properties, `locals.foo = 'bar'`

File: src/content/docs/en/guides/middleware.mdx [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+## API reference
+
+### onRequest()
+
+A required exported function from `src/middleware.js` that will be called before rendering every page or API route. It accepts two arguments: [context](#context) and [next()](#next)
+
+### context
+
+An object that includes information to be made available to other middleware, API routes and `.astro` routes during the rendering process.
```
Comment (matthewp) [inline src/content/docs/en/guides/middleware.mdx:153]: This is the same object that is passed to API endpoints. Not sure if that's worth mentioning here or not.

**Review by ematipico — APPROVED** [review]

File: src/content/docs/en/guides/middleware.mdx [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+This is a required argument passed to `onRequest()` that must contain the [`locals`](#locals) object as well as any additional properties to be shared during rendering. For example, the `context` object may include cookies used in authentication.
+
+### next ()
+
+A function that retrieves that intercepts (reads and modifies) the `Response` of a `Request` or calls the "next" middleware in the chain. For example, this function could modify the HTML body of a response.
+
+This is an optional argument of `onRequest()`.
+
+
+### locals
```
Comment (ematipico) [inline src/content/docs/en/guides/middleware.mdx:164]: Do you think it's worth mentioning that Astro will make these checks **only** in **dev** mode?

File: src/content/docs/en/guides/middleware.mdx [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+### onRequest()
+
+A required exported function from `src/middleware.js` that will be called before rendering every page or API route. It accepts two arguments: [context](#context) and [next()](#next)
+
+### context
+
+An object that includes information to be made available to other middleware, API routes and `.astro` routes during the rendering process. 
+
+This is a required argument passed to `onRequest()` that must contain the [`locals`](#locals) object as well as any additional properties to be shared during rendering. For example, the `context` object may include cookies used in authentication.
```
[the remark this answers is not in the document]
Comment (sarah11918) [reply src/content/docs/en/guides/middleware.mdx:155]: Thanks, I wasn't *sure* whether this was required, but the best way to find out is to write that it is and let others correct me!  :smile:

File: src/content/docs/en/guides/middleware.mdx [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+A function that intercepts (reads and modifies) the `Response` of a `Request` or calls the "next" middleware in the chain and returns a `Response`. For example, this function could modify the HTML body of a response.
+
+This is an optional argument of `onRequest()`; however, your middleware **must** either return a `Response` or call `next()`.
+
+
+### `locals`
+An object containing data from a response that can be manipulated inside middleware. 
+9
+This `locals` object is forwarded across the request handling process and is available as a property to APIContext and AstroGlobal. This allows data to be shared between middlewares, API routes, and .astro pages. This is useful for storing request-specific data, such as user data, across the rendering step.
```
Comment (ematipico) [inline src/content/docs/en/guides/middleware.mdx:158]: FYI: currently, we disabled the middleware functionality for API routes ([endpoints](https://docs.astro.build/en/core-concepts/endpoints/#static-file-endpoints)?) that use the following pattern:

```js
export function get({}) {
	return {
		body: "some text"
	}
}
```

I mean, those endpoints that only return an object that contains `body`. Astro will log a warning then that happens.

**Review by ematipico — APPROVED** [review]

**Pull request merged**